File: dynamic-programming/MinimumCoverage.java
/***
Given an integer array of size n
Find the minimum number of consecutive elements in the array that 
contain all the values of that array
*/

import java.util.Map;
import java.util.HashMap;
import java.util.Set;
import java.util.HashSet;

public class MinimumCoverage {
	
	private static final int[] A = new int[]{1, 3, 0, 7, 4, 3, 5, 4, 1, 9, 7}; 
	
	public static void main(String[] args) {
		System.out.println(findMinimumCoverage(A));
	}
	
	public static int findMinimumCoverage(int[] a) {
		
		int n = a.length;
		
		Map<Integer, Set<Integer>> coverageMap =  new HashMap<Integer, Set<Integer>>();
		
		Set<Integer> allValues =  new HashSet<Integer>();
		
		for (int i = 0; i < n; i++) {
			allValues.add(a[i]);
			Set<Integer> s = new HashSet<Integer>();
			s.add(a[i]);
			coverageMap.put(i*n+i, s);
		}
		
		int valueCount = allValues.size();
		int minCoverageRange = n;
		
		for (int i = 0; i < n-1; i++) {
			for (int j =i + 1; j < n; j++) {
				Set<Integer> s = new HashSet<Integer>();
			    s.addAll(coverageMap.get(i*n+j-1));	
				s.add(a[j]);
				coverageMap.put(i*n+j, s);
				if (s.size() == valueCount && j-i+1 < minCoverageRange) {
					minCoverageRange = j-i+1;
				}
			}
		}	 
		
		return minCoverageRange;
	}
}

File: dynamic-programming/minimum_coverage.py
# Given an integer list of size n
# Find the minimum number of consecutive elements in the array that
# contain all the values of that array


def find_minimum_coverage(numbers):
    n = len(numbers);
    
    s = set(numbers)
    value_count = len(s)
    
    coverage_dict = {
        (i,i):  set([numbers[i]]) for i in range(n)   
    }
    
    min_coverage = n;
    
    for i in range(n-1):
        for j in range(i+1, n):
            coverage_dict.update({
                (i, j): set(coverage_dict[(i, j-1)])
            })
            coverage_dict[(i, j)].add(numbers[j])
            if len(coverage_dict[(i, j)]) == value_count and j-i+1 < min_coverage:
                min_coverage = j-i+1 
            
    return min_coverage
    
    
if __name__=='__main__':
    print find_minimum_coverage([1, 3, 0, 7, 4, 3, 5, 4, 1, 9, 7])    
    
